VirtoCommerce.CartModule.Data/Model/ShoppingCartEntity.cs

```diff
@@ -176,14 +175,20 @@ public virtual ShoppingCartEntity FromModel(ShoppingCart cart, PrimaryKeyResolvi
             {
                 TaxDetails = new ObservableCollection<TaxDetailEntity>(cart.TaxDetails.Select(x => AbstractTypeFactory<TaxDetailEntity>.TryCreateInstance().FromModel(x)));
             }
-            //back compatibility changes
+
+            var coupons = new List<CouponEntity>();
+
             if (cart.Coupon != null)
             {
-                Coupons = new ObservableCollection<CouponEntity>(new[] { new CouponEntity { Code = cart.Coupon } });
+                coupons.Add(new CouponEntity { Code = cart.Coupon });
             }
             if (cart.Coupons != null)
             {
-                Coupons = new ObservableCollection<CouponEntity>(cart.Coupons.Select(x => new CouponEntity { Code = x }));
+                coupons = cart.Coupons.Select(x => new CouponEntity { Code = x }).Concat(coupons).Distinct().ToList();
```

Comment:
Owerwrites the previously added coupon. Use `else` instead.

Reply:
To prevent overriding, .Concat(coupons) is used.

Reply:
Renamed coupons local variable to avoid mix-up.